src/compiler/crystal/codegen/llvm_builder_helper.cr

```diff
@@ -189,14 +185,31 @@ module Crystal
       end
     end
 
-    def cast_to(value, type)
-      bit_cast value, llvm_type(type)
+    def cast_to(value : LLVM::ValueMethods, type : Type)
+      pointer_cast value, llvm_type(type)
+    end
+
+    def cast_to_pointer(value : LLVM::ValueMethods, type : Type)
+      pointer_cast value, llvm_type(type).pointer
     end
 
-    def cast_to_pointer(value, type)
-      bit_cast value, llvm_type(type).pointer
+    def cast_to_void_pointer(pointer : LLVM::ValueMethods)
+      pointer_cast pointer, llvm_context.void_pointer
     end
 
+    # *type* must be a pointer type; on LLVM 15.0 or above *type* is not
+    # evaluated at all and *value* is returned unchanged, because all opaque
+    # pointer types (in the same context) are identical
+    {% if LibLLVM::IS_LT_150 %}
+      def pointer_cast(value : LLVM::ValueMethods, type : LLVM::Type)
+        bit_cast(value, type)
+      end
+    {% else %}
+      macro pointer_cast(value, type)
+        \{{ value }}
+      end
```

Comment:
Why is this a macro? Feels very odd to have the same call turn out to be a macro or method call depending on the LLVM version.